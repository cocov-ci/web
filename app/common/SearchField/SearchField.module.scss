@import "styles/variables";

.searchField {
  position: relative;

  input {
    border-radius: 8px;
    width: 208px;
    border: 1px solid $colorGray;
    background-color: $colorGrayLight;
    font-family: var(--satoshi-font);
    font-size: 17px;
    padding: 7px 10px 7px 33px;
  }

  .icon {
    position: absolute;
    text-align: left;
    top: 50%;
    left: 8px;
    display: inline-flex;
    transform: translateY(-50%);
  }
}
