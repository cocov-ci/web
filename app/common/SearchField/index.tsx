'use client'

import classNames from 'classnames'
import { LucideSearch } from 'lucide-react'
import React from 'react'

import Input from 'app/common/Input'

import styles from './SearchField.module.scss'

interface SearchFieldProps {
  className?: string
  loading?: boolean
  onSearch: (args: string) => void
}

const SearchField = ({
  className,
  loading = false,
  onSearch,
}: SearchFieldProps) => {
  let timer: ReturnType<typeof setTimeout>

  const SearchDebounce = (e: React.ChangeEvent<HTMLInputElement>) => {
    const value = e.target.value

    if (timer) {
      clearTimeout(timer)
    }

    timer = setTimeout(() => {
      onSearch(value)
    }, 500)
  }

  return (
    <div className={classNames(styles.searchField, className)}>
      <Input
        className={styles.input}
        icon={LucideSearch}
        loading={loading}
        onChange={SearchDebounce}
        placeholder="Type to Search"
        type="text"
        width="208px"
      />
    </div>
  )
}

export default SearchField
