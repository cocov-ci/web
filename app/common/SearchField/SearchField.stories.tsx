import type { Meta, StoryObj } from '@storybook/react'

import SearchField from './index'

const meta: Meta<typeof SearchField> = {
  title: 'Common/SearchField',
  component: SearchField,
  tags: ['autodocs'],
}

export default meta

type Story = StoryObj<typeof SearchField>

export const Default: Story = {
  args: {
    className: '',
    loading: false,
  },
}

export const Loading: Story = {
  args: {
    loading: true,
  },
}
