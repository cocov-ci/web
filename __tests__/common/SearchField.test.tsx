import { fireEvent, render, screen } from '@testing-library/react'

import SearchField from 'app/common/SearchField'

import '@testing-library/jest-dom'

describe('common/SearchField', () => {
  const { container } = render(<SearchField onSearch={() => null} />)

  it('renders SearchField component', () => {
    expect(screen.getByPlaceholderText('Type to Search')).toBeVisible()
    expect(container.querySelector('svg')).toBeVisible()
  })

  it('renders SearchField component with "loading" props', () => {
    const { container } = render(
      <SearchField loading={true} onSearch={() => null} />,
    )

    expect(screen.getByPlaceholderText('Type to Search')).toBeVisible()
    expect(container.querySelector('svg')).toBeVisible()
    expect(container.querySelector('img')).toBeVisible()
  })

  it('renders SearchField component with "disabled" props', () => {
    const { container } = render(
      <SearchField disabled={true} onSearch={() => null} />,
    )

    expect(screen.getByPlaceholderText('Type to Search')).toBeVisible()
    expect(container.querySelector('svg')).toBeVisible()
    expect(container.querySelector('input')).toHaveAttribute('disabled')
  })

  it('renders SearchField component with "onSearch" props', () => {
    const handleOnSearch = jest.fn()
    render(<SearchField disabled={true} onSearch={handleOnSearch} />)

    const input = screen.getByRole('textbox') as HTMLInputElement

    fireEvent.change(input, { target: { value: 'text her' } })
    fireEvent.change(input, { target: { value: 'text here...' } })

    setTimeout(() => {
      expect(handleOnSearch).toHaveBeenCalledTimes(1)
      expect(input.value).toBe('text here...')
    }, 500)
  })

  it('renders SearchField snapshots', () => {
    expect(container).toMatchSnapshot()
  })
})
