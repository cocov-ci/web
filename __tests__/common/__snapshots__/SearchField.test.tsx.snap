// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`common/SearchField renders SearchField snapshots 1`] = `<div />`;
